src/shogun/base/AnyParameter.h

```diff
@@ -28,101 +28,85 @@ namespace shogun
 		GRADIENT_AVAILABLE = 1
 	};
 	/** @brief Class AnyParameterProperties keeps track of parameter properties.
-	 *
-	 * Currently there are three types of parameters:
-	 * -# HYPERPARAMETER: the parameter determines how the training is
-	 * performed, e.g. regularisation
-	 * -# GRADIENT_PARAM: the parameter is used in gradient updates
-	 * -# MODEL_PARAM: model parameter that is trained, e.g. weight and bias
-	 * vectors and bias
+	 * The parameter properties can be either true or false.
+	 * These properties describe if a parameter is for example a hyperparameter
+	 * or if it has a gradient.
 	 */
 	class AnyParameterProperties
 	{
 	public:
-		static const int32_t HYPERPARAMETER = 0x00000001;
-		static const int32_t GRADIENT_PARAM = 0x00000010;
-		static const int32_t MODEL_PARAM = 0x00000100;
+		static const int32_t HYPER = 1u << 0;
+		static const int32_t GRADIENT = 1u << 1;
+		static const int32_t MODEL = 1u << 2;
 
-		/** default constructor where all parameter types are switched to false
+		/** Default constructor where all parameter properties are false
 		 */
-		AnyParameterProperties() : m_description(), m_mask_attribute(0x00000000)
-		{
-		}
-		/** legacy constructor */
+		AnyParameterProperties()
+		    : m_description("No description given"), m_attribute_mask(0)
+		{
+		}
+		/** Constructor
+		 * @param description parameter description
+		 * @param hyperparameter set to true for parameters that determine
+		 * how training is performed, e.g. regularisation parameters
+		 * @param gradient set to true for parameters required for gradient
+		 * updates
+		 * @param model set to true for parameters used in inference, e.g.
+		 * weights and bias
+		 * */
 		AnyParameterProperties(
 		    std::string description,
-		    EModelSelectionAvailability model_selection = MS_NOT_AVAILABLE,
+		    EModelSelectionAvailability hyperparameter = MS_NOT_AVAILABLE,
 		    EGradientAvailability gradient = GRADIENT_NOT_AVAILABLE,
-		    bool hyperparameter = false)
-		    : m_description(description), m_model_selection(model_selection),
+		    bool model = false)
+		    : m_description(description), m_model_selection(hyperparameter),
 		      m_gradient(gradient)
 		{
-			m_mask_attribute = 0x00000000;
-			if (model_selection)
-				m_mask_attribute |= MODEL_PARAM;
-			if (gradient)
-				m_mask_attribute |= GRADIENT_PARAM;
-			if (hyperparameter)
-				m_mask_attribute |= HYPERPARAMETER;
+			m_attribute_mask = (hyperparameter << 0 & HYPER) |
```

Comment:
I was more thinking
`(1 << HYPER) & model_selection | (1 << GRADIENT) && gradient` etc

Reply:
Oh that's too low-level. Can we introduce a constexpr function for that?

Reply:
yeah actually good point! :)